bap-vibes/src/arm_selector.ml

```diff
@@ -368,6 +381,18 @@ module ARM_ops = struct
     | n when n <= 0xFFFF -> Ops.movw
     | _ -> Ops.ldr 
 
+  (* Detect the movcc pattern when storing the result of a comparison. *)
+  let movcc_pattern (o1 : Ir.operation) (o2 : Ir.operation) : bool =
```

Comment:
Example:
```
movgt temp1, #1
movle temp2, #0
mov dst, temp1
```

`temp1` and `temp2` are congruent, however the optimization in `arm_mov` was producing the following code:

 ```
movgt temp1, #1
movle dst, #0
```

We recorded the congruence already, so when we try to ship it off to Minizinc we get an error since `temp2` no longer exists.

Reply:
Additionally, we can't propagate `dst` to the LHS of both conditional moves, since the program would no longer be in linear SSA form.

Reply:
I don't agree that cmp depends on temp1. I think 
```
temp1 := mov #1
ZF := cmp x, y
dst := movle temp1 #0 ZF
```
is my suggested functional model of the siutation.

Reply:
Otherwise, we would be forced to always use the 32-bit encoding of `mov`, thus we miss an opportunity to make the patch smaller.

Reply:
What is compiled to `movs`? The assembler is going to turn multiple instructions into a movs?

Reply:
Depending on whether we're on Thumb or ARM mode, the `Ops.mov` function will either return the `movs` or `mov` opcode. The reason for this is, again, to use the smallest encoding possible.

Reply:
Note that before the cp6 PR, we were not using the `.syntax unified` directive for the assembler, so this distinction was not necessary. However, the assembler was secretly choosing `movs` where appropriate.

Reply:
If we're compiling to Thumb, then the semantics is:
```
temp1, ZF, CF, ... := movs #1
ZF , CF, ... := cmp x, y
dst := movle temp1 #0 ZF CF
```

Reply:
Ok that sounds good. I'm in favor of just modelling the flags rather than using fake void registers to force ordering when we can. It's conceptually clearer to me.

Reply:
See here: https://www.keil.com/support/man/docs/armasm/armasm_dom1361289878994.htm

In order to use the 16-bit encoding of `mov`  with an immediate operand on Thumb, we have to either be inside an IT block or explicitly set the flags according to the result.

Reply:
Temps need to be uniquely defined, so some of your "ZF" and "CF" are actually different temps. I believe the anti clobbering behavior of the "reg_t" ZF will enforce the correct ordering here

Reply:
Plus, we don't seem to be modeling flags in the model either, so maybe this could be a future PR?

Reply:
Yes, It should. We should move off of "void" nonsense as much as possible. It's confusing